src/ray/pubsub/subscriber.cc

```diff
@@ -18,174 +18,272 @@ namespace ray {
+template <typename MessageID>
+inline absl::optional<SubscriptionCallback>
+SubscriberChannel<MessageID>::GetSubscriptionCallback(
+    const rpc::Address &publisher_address, const MessageID &message_id) const {
   const auto publisher_id = PublisherID::FromBinary(publisher_address.worker_id());
   auto subscription_it = subscription_map_.find(publisher_id);
   if (subscription_it == subscription_map_.end()) {
     return absl::nullopt;
   }
-  auto callback_it = subscription_it->second.subscription_callback_map_.find(object_id);
-  bool exist = callback_it != subscription_it->second.subscription_callback_map_.end();
+  auto callback_it = subscription_it->second.subscription_callback_map.find(message_id);
+  bool exist = callback_it != subscription_it->second.subscription_callback_map.end();
   if (!exist) {
     return absl::nullopt;
   }
   auto subscription_callback = callback_it->second.first;
   return absl::optional<SubscriptionCallback>{subscription_callback};
 }
 
-inline absl::optional<SubscriptionCallback> Subscriber::GetFailureCallback(
-    const rpc::Address &publisher_address, const ObjectID &object_id) const {
+template <typename MessageID>
+inline absl::optional<SubscriptionFailureCallback>
+SubscriberChannel<MessageID>::GetFailureCallback(const rpc::Address &publisher_address,
+                                                 const MessageID &message_id) const {
   const auto publisher_id = PublisherID::FromBinary(publisher_address.worker_id());
   auto subscription_it = subscription_map_.find(publisher_id);
   if (subscription_it == subscription_map_.end()) {
     return absl::nullopt;
   }
-  auto callback_it = subscription_it->second.subscription_callback_map_.find(object_id);
-  bool exist = callback_it != subscription_it->second.subscription_callback_map_.end();
+  auto callback_it = subscription_it->second.subscription_callback_map.find(message_id);
+  bool exist = callback_it != subscription_it->second.subscription_callback_map.end();
   if (!exist) {
     return absl::nullopt;
   }
-  auto subscription_callback = callback_it->second.second;
-  return absl::optional<SubscriptionCallback>{subscription_callback};
+  auto subscription_failure_callback = callback_it->second.second;
+  return absl::optional<SubscriptionFailureCallback>{subscription_failure_callback};
 }
 
-bool Subscriber::AssertNoLeak() const {
-  for (const auto &subscription : subscription_map_) {
-    if (subscription.second.subscription_callback_map_.size() != 0) {
-      return false;
+template <typename MessageID>
+bool SubscriberChannel<MessageID>::SubscriptionExists(const PublisherID &publisher_id) {
+  return subscription_map_.count(publisher_id);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/// Subscriber
+///////////////////////////////////////////////////////////////////////////////
+
+inline std::shared_ptr<SubscribeChannelInterface> Subscriber::Channel(
```

Comment:
why inline here?

Reply:
because this function is frequently called, but pretty tiny.

Reply:
Looks like inline doesn't mean the function is actually inlined. I will move the definition to the header instead to get true inline functions.